# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/cookiestore/CookieStoreNotifier.cpp
# Commit: 0d57b435d872
# Full Hash: 0d57b435d872751c2a26879fb04bf8745314e37a
# Author: Andrea Marchesini <amarchesini@mozilla.com>
# Date: 2024-09-12 09:23:07
# Regressor Bug: 1475599
# File Overlap Count: 1
# Description:
#   Bug 1475599 - part 8 - CookieStore API - e10s support, r=smaug,cookie-reviewers,timhuang
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D217574
# ==============================================================================

diff -r a63a5bcb0524 -r 0d57b435d872 dom/cookiestore/CookieStoreNotifier.cpp
--- a/dom/cookiestore/CookieStoreNotifier.cpp	Wed Sep 11 16:26:57 2024 +0000
+++ b/dom/cookiestore/CookieStoreNotifier.cpp	Wed Sep 11 16:26:57 2024 +0000
@@ -6,6 +6,7 @@
 
 #include "CookieStoreNotifier.h"
 #include "CookieStore.h"
+#include "CookieChangeEvent.h"
 #include "mozilla/net/CookieCommons.h"
 #include "mozilla/dom/Document.h"
 #include "mozilla/dom/WorkerPrivate.h"
@@ -192,43 +193,53 @@
 
   bool deletedEvent = action == nsICookieNotification::COOKIE_DELETED;
 
-  mEventTarget->Dispatch(NS_NewRunnableFunction(
-      __func__, [self = RefPtr(this), item, deletedEvent] {
-        if (!self->mCookieStore) {
-          return;
-        }
+  if (mEventTarget->IsOnCurrentThread()) {
+    DispatchEvent(item, deletedEvent);
+  } else {
+    mEventTarget->Dispatch(NS_NewRunnableFunction(
+        __func__, [self = RefPtr(this), item, deletedEvent] {
+          self->DispatchEvent(item, deletedEvent);
+        }));
+  }
 
-        RefPtr<Event> event = deletedEvent
-                                  ? CookieChangeEvent::CreateForDeletedCookie(
-                                        self->mCookieStore, item)
-                                  : CookieChangeEvent::CreateForChangedCookie(
-                                        self->mCookieStore, item);
+  return NS_OK;
+}
 
-        if (!event) {
-          return;
-        }
+void CookieStoreNotifier::DispatchEvent(const CookieListItem& aItem,
+                                        bool aDeletedEvent) {
+  MOZ_ASSERT(mEventTarget->IsOnCurrentThread());
+
+  if (!mCookieStore) {
+    return;
+  }
 
-        if (NS_IsMainThread()) {
-          nsCOMPtr<nsPIDOMWindowInner> window =
-              self->mCookieStore->GetOwnerWindow();
-          if (!window) {
-            return;
-          }
+  RefPtr<Event> event =
+      aDeletedEvent
+          ? CookieChangeEvent::CreateForDeletedCookie(mCookieStore, aItem)
+          : CookieChangeEvent::CreateForChangedCookie(mCookieStore, aItem);
+
+  if (!event) {
+    return;
+  }
 
-          RefPtr<BrowsingContext> bc = window->GetBrowsingContext();
-          if (!bc) {
-            return;
-          }
+  if (NS_IsMainThread()) {
+    nsCOMPtr<nsPIDOMWindowInner> window = mCookieStore->GetOwnerWindow();
+    if (!window) {
+      return;
+    }
 
-          if (bc->IsInBFCache()) {
-            self->mDelayedDOMEvents.AppendElement(event);
-            return;
-          }
-        }
+    RefPtr<BrowsingContext> bc = window->GetBrowsingContext();
+    if (!bc) {
+      return;
+    }
 
-        self->mCookieStore->DispatchEvent(*event);
-      }));
-  return NS_OK;
+    if (bc->IsInBFCache()) {
+      mDelayedDOMEvents.AppendElement(event);
+      return;
+    }
+  }
+
+  mCookieStore->DispatchEvent(*event);
 }
 
 void CookieStoreNotifier::FireDelayedDOMEvents() {